src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -1735,5 +1759,87 @@ private RendezvousAwaitable<int> InternalReceiveResponseHeadersAsync(WinHttpRequ
 
             return state.LifecycleAwaitable;
         }
+
+        internal bool GetCertificateFromCache(IPAddress ipAddress, [MaybeNullWhen(false)] out byte[] rawCertificateBytes)
+        {
+            if (_cachedCertificates.TryGetValue(ipAddress, out CachedCertificateValue? cachedValue))
+            {
+                cachedValue.LastUsedTime = Stopwatch.GetTimestamp();
+                rawCertificateBytes = cachedValue.RawCertificateData;
+                return true;
+            }
+
+            rawCertificateBytes = null;
+            return false;
+        }
+
+        internal bool TryAddCertificateToCache(IPAddress address, byte[] rawCertificateData)
```

Comment:
```suggestion
        internal void AddCertificateToCache(IPAddress address, byte[] rawCertificateData)
```
Nit: the one usage of this ignores the return value